### PR TITLE
Stage4

### PR DESCRIPTION
**Pull request by Valaubr**

### CODE REVIEW

**Comment by Valaubr** [comment]
### 1. PUT
Скрипт нагрузки: wrk -t4 -c100 -d20m -R3000 -s /home/valaubr/GitDirs/2020-highload-dht/wrkAndProfile/scripts/put.lua --latency http://127.0.0.1:8080

Running 20m test @ http://127.0.0.1:8080
  4 threads and 100 connections
  Thread calibration: mean lat.: 1.111ms, rate sampling interval: 10ms
  Thread calibration: mean lat.: 1.113ms, rate sampling interval: 10ms
  Thread calibration: mean lat.: 1.101ms, rate sampling interval: 10ms
  Thread calibration: mean lat.: 1.105ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.11ms  603.31us  39.84ms   82.12%
    Req/Sec   791.68    111.86     2.11k    79.98%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.08ms
 75.000%    1.39ms
 90.000%    1.67ms
 99.000%    2.32ms
 99.900%    8.28ms
 99.990%   13.25ms
 99.999%   23.12ms
100.000%   39.87ms

  Detailed Percentile spectrum:
       Value   Percentile   TotalCount 1/(1-Percentile)

       0.065     0.000000            1         1.00
       0.520     0.100000       358301         1.11
       0.683     0.200000       714599         1.25
       0.825     0.300000      1072546         1.43
       0.955     0.400000      1429591         1.67
       1.078     0.500000      1784940         2.00
       1.139     0.550000      1965595         2.22
       1.198     0.600000      2142435         2.50
       1.258     0.650000      2321379         2.86
       1.320     0.700000      2499183         3.33
       1.387     0.750000      2679109         4.00
       1.422     0.775000      2766501         4.44
       1.461     0.800000      2855831         5.00
       1.504     0.825000      2945145         5.71
       1.553     0.850000      3035369         6.67
       1.608     0.875000      3124526         8.00
       1.638     0.887500      3168140         8.89
       1.671     0.900000      3212723        10.00
       1.708     0.912500      3257498        11.43
       1.749     0.925000      3302466        13.33
       1.796     0.937500      3346742        16.00
       1.823     0.943750      3369364        17.78
       1.853     0.950000      3391809        20.00
       1.886     0.956250      3413788        22.86
       1.925     0.962500      3435851        26.67
       1.973     0.968750      3458350        32.00
       2.000     0.971875      3469239        35.56
       2.031     0.975000      3480378        40.00
       2.067     0.978125      3491769        45.71
       2.109     0.981250      3502986        53.33
       2.161     0.984375      3513957        64.00
       2.195     0.985938      3519593        71.11
       2.233     0.987500      3525110        80.00
       2.281     0.989062      3530582        91.43
       2.349     0.990625      3536165       106.67
       2.463     0.992188      3541749       128.00
       2.569     0.992969      3544525       142.22
       2.775     0.993750      3547310       160.00
       3.107     0.994531      3550097       182.86
       3.513     0.995313      3552890       213.33
       4.045     0.996094      3555681       256.00
       4.375     0.996484      3557081       284.44
       4.731     0.996875      3558477       320.00
       5.155     0.997266      3559862       365.71
       5.683     0.997656      3561261       426.67
       6.259     0.998047      3562656       512.00
       6.587     0.998242      3563346       568.89
       6.955     0.998437      3564046       640.00
       7.375     0.998633      3564744       731.43
       7.815     0.998828      3565437       853.33
       8.343     0.999023      3566136      1024.00
       8.615     0.999121      3566488      1137.78
       8.911     0.999219      3566836      1280.00
       9.191     0.999316      3567178      1462.86
       9.559     0.999414      3567532      1706.67
       9.919     0.999512      3567878      2048.00
      10.127     0.999561      3568051      2275.56
      10.399     0.999609      3568229      2560.00
      10.671     0.999658      3568404      2925.71
      10.951     0.999707      3568575      3413.33
      11.287     0.999756      3568749      4096.00
      11.471     0.999780      3568836      4551.11
      11.711     0.999805      3568922      5120.00
      11.991     0.999829      3569013      5851.43
      12.279     0.999854      3569100      6826.67
      12.711     0.999878      3569183      8192.00
      13.039     0.999890      3569226      9102.22
      13.295     0.999902      3569270     10240.00
      13.655     0.999915      3569313     11702.86
      14.199     0.999927      3569357     13653.33
      14.823     0.999939      3569401     16384.00
      15.167     0.999945      3569423     18204.44
      15.623     0.999951      3569444     20480.00
      16.159     0.999957      3569466     23405.71
      17.071     0.999963      3569488     27306.67
      17.791     0.999969      3569510     32768.00
      18.159     0.999973      3569520     36408.89
      18.767     0.999976      3569531     40960.00
      19.375     0.999979      3569542     46811.43
      20.271     0.999982      3569553     54613.33
      21.567     0.999985      3569564     65536.00
      21.919     0.999986      3569569     72817.78
      22.383     0.999988      3569575     81920.00
      22.687     0.999989      3569580     93622.86
      23.439     0.999991      3569586    109226.67
      24.095     0.999992      3569591    131072.00
      24.591     0.999993      3569594    145635.56
      24.863     0.999994      3569597    163840.00
      25.103     0.999995      3569599    187245.71
      26.047     0.999995      3569602    218453.33
      26.319     0.999996      3569605    262144.00
      27.151     0.999997      3569606    291271.11
      29.439     0.999997      3569608    327680.00
      29.679     0.999997      3569609    374491.43
      30.159     0.999998      3569610    436906.67
      31.167     0.999998      3569612    524288.00
      31.167     0.999998      3569612    582542.22
      31.199     0.999998      3569613    655360.00
      33.503     0.999999      3569614    748982.86
      33.503     0.999999      3569614    873813.33
      35.967     0.999999      3569615   1048576.00
      35.967     0.999999      3569615   1165084.44
      36.351     0.999999      3569616   1310720.00
      36.351     0.999999      3569616   1497965.71
      36.351     0.999999      3569616   1747626.67
      36.479     1.000000      3569617   2097152.00
      36.479     1.000000      3569617   2330168.89
      36.479     1.000000      3569617   2621440.00
      36.479     1.000000      3569617   2995931.43
      36.479     1.000000      3569617   3495253.33
      39.871     1.000000      3569618   4194304.00
      39.871     1.000000      3569618          inf
#[Mean    =        1.113, StdDeviation   =        0.603]
#[Max     =       39.840, Total count    =      3569618]
#[Buckets =           27, SubBuckets     =         2048]
----------------------------------------------------------
  3599863 requests in 20.00m, 284.95MB read
Requests/sec:   2999.89
Transfer/sec:    243.15KB

**CPU**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/cpu_put.svg

Видно что что проксирование занимает 31.66% вставка же 3.82%

**Alloc**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/alloc_put.svg

Аллокация памяти в основном происходит во время проксирования - 49.77%

**Lock**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/lock_put.svg

Почти не отличается от прошлых результатов:
Селекторы блокируются, что отдать задачу в пул потоков.
Воркеры - запись данных в хранилище.

### 2.GET

Скрипт нагрузки: wrk -t4 -c100 -d20m -R3000 -s /home/valaubr/GitDirs/2020-highload-dht/wrkAndProfile/scripts/get.lua --latency http://127.0.0.1:8080
Running 20m test @ http://127.0.0.1:8080
  4 threads and 100 connections
  Thread calibration: mean lat.: 81.207ms, rate sampling interval: 628ms
  Thread calibration: mean lat.: 14.568ms, rate sampling interval: 69ms
  Thread calibration: mean lat.: 85.918ms, rate sampling interval: 703ms
  Thread calibration: mean lat.: 64.366ms, rate sampling interval: 538ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.71ms    1.23ms  56.00ms   93.24%
    Req/Sec   200.94     39.86   284.00     77.58%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.48ms
 75.000%    1.82ms
 90.000%    2.64ms
 99.000%    5.93ms
 99.900%   16.53ms
 99.990%   30.45ms
 99.999%   44.67ms
100.000%   56.03ms

  Detailed Percentile spectrum:
       Value   Percentile   TotalCount 1/(1-Percentile)

       0.618     0.000000            1         1.00
       0.930     0.100000       114344         1.11
       1.092     0.200000       229331         1.25
       1.204     0.300000       343526         1.43
       1.344     0.400000       457167         1.67
       1.484     0.500000       571735         2.00
       1.536     0.550000       628516         2.22
       1.599     0.600000       685879         2.50
       1.668     0.650000       742877         2.86
       1.741     0.700000       800107         3.33
       1.822     0.750000       856912         4.00
       1.900     0.775000       885414         4.44
       2.055     0.800000       914061         5.00
       2.199     0.825000       942471         5.71
       2.327     0.850000       971015         6.67
       2.469     0.875000       999588         8.00
       2.549     0.887500      1013817         8.89
       2.639     0.900000      1028260        10.00
       2.747     0.912500      1042353        11.43
       2.867     0.925000      1056725        13.33
       2.991     0.937500      1070997        16.00
       3.059     0.943750      1078097        17.78
       3.137     0.950000      1085257        20.00
       3.227     0.956250      1092341        22.86
       3.335     0.962500      1099480        26.67
       3.487     0.968750      1106631        32.00
       3.579     0.971875      1110158        35.56
       3.693     0.975000      1113729        40.00
       3.835     0.978125      1117344        45.71
       4.025     0.981250      1120869        53.33
       4.383     0.984375      1124431        64.00
       4.651     0.985938      1126214        71.11
       5.047     0.987500      1128005        80.00
       5.583     0.989062      1129791        91.43
       6.199     0.990625      1131578       106.67
       6.947     0.992188      1133355       128.00
       7.371     0.992969      1134251       142.22
       7.859     0.993750      1135148       160.00
       8.423     0.994531      1136037       182.86
       9.111     0.995313      1136924       213.33
      10.007     0.996094      1137823       256.00
      10.487     0.996484      1138266       284.44
      11.039     0.996875      1138711       320.00
      11.671     0.997266      1139155       365.71
      12.383     0.997656      1139605       426.67
      13.167     0.998047      1140049       512.00
      13.655     0.998242      1140270       568.89
      14.343     0.998437      1140495       640.00
      14.943     0.998633      1140716       731.43
      15.767     0.998828      1140940       853.33
      16.623     0.999023      1141163      1024.00
      17.119     0.999121      1141275      1137.78
      17.711     0.999219      1141385      1280.00
      18.543     0.999316      1141498      1462.86
      19.503     0.999414      1141609      1706.67
      20.495     0.999512      1141720      2048.00
      21.103     0.999561      1141776      2275.56
      21.999     0.999609      1141831      2560.00
      22.719     0.999658      1141888      2925.71
      23.679     0.999707      1141943      3413.33
      24.959     0.999756      1141999      4096.00
      25.455     0.999780      1142028      4551.11
      26.191     0.999805      1142056      5120.00
      27.199     0.999829      1142082      5851.43
      28.207     0.999854      1142110      6826.67
      29.167     0.999878      1142138      8192.00
      29.903     0.999890      1142152      9102.22
      30.559     0.999902      1142167     10240.00
      31.535     0.999915      1142180     11702.86
      32.479     0.999927      1142194     13653.33
      33.279     0.999939      1142208     16384.00
      33.823     0.999945      1142215     18204.44
      35.487     0.999951      1142223     20480.00
      36.831     0.999957      1142229     23405.71
      38.975     0.999963      1142236     27306.67
      39.711     0.999969      1142243     32768.00
      40.159     0.999973      1142246     36408.89
      41.151     0.999976      1142250     40960.00
      42.143     0.999979      1142253     46811.43
      42.687     0.999982      1142257     54613.33
      42.879     0.999985      1142260     65536.00
      44.191     0.999986      1142262     72817.78
      44.287     0.999988      1142264     81920.00
      44.415     0.999989      1142265     93622.86
      45.119     0.999991      1142267    109226.67
      47.263     0.999992      1142269    131072.00
      48.383     0.999993      1142270    145635.56
      50.079     0.999994      1142272    163840.00
      50.079     0.999995      1142272    187245.71
      50.079     0.999995      1142272    218453.33
      51.103     0.999996      1142273    262144.00
      52.287     0.999997      1142275    291271.11
      52.287     0.999997      1142275    327680.00
      52.287     0.999997      1142275    374491.43
      52.287     0.999998      1142275    436906.67
      52.287     0.999998      1142275    524288.00
      52.479     0.999998      1142276    582542.22
      52.479     0.999998      1142276    655360.00
      52.479     0.999999      1142276    748982.86
      52.479     0.999999      1142276    873813.33
      52.479     0.999999      1142276   1048576.00
      56.031     0.999999      1142277   1165084.44
      56.031     1.000000      1142277          inf
#[Mean    =        1.707, StdDeviation   =        1.230]
#[Max     =       56.000, Total count    =      1142277]
#[Buckets =           27, SubBuckets     =         2048]
----------------------------------------------------------
  1152232 requests in 20.00m, 92.30MB read
  Socket errors: connect 0, read 0, write 0, timeout 40672
Requests/sec:    960.19
Transfer/sec:     78.76KB 

**CPU**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/cpu_get.svg

Как видно, в отличии от вставки, получение денных отъедает существенно большую часть CPU по сравнению с проксированием: 84.99% к 4.61%, так же стоит отметить что сложнейшая операция в DAO.get - получение итератора.  

**Alloc**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/alloc_get.svg

Проксирование - 13.27%.

**Lock**
https://github.com/Valaubr/2020-highload-dht/blob/stage4/wrkAndProfile/stage4/lock_get.svg

По аналогии с PUT

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        }
+    }
+
+    private boolean checkId(@NotNull final String id, @NotNull final HttpSession session) {
+        if (id.strip().isEmpty()) {
+            try {
+                session.sendResponse(new Response(Response.BAD_REQUEST, Response.EMPTY));
+                return true;
+            } catch (IOException e) {
+                logger.error(UNIVERSAL_MESSAGE, e);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:239]: Сообщения логгера обычно не выносят в константу (по крайней мере, если это сообщение 1 раз встречается).

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    private void proxy(@NotNull final String node,
+                       @NotNull final Request request,
+                       @NotNull final HttpSession session) {
+        try {
+            request.addHeader("X-Proxy-For: " + node);
+            session.sendResponse(nodeToClient.get(node).invoke(request));
+        } catch (IOException | InterruptedException | PoolException | HttpException e) {
+            logger.error("Can`t proxy request: ", e);
+            messageOfInternalError(session, "Socket go to sleep");
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:229]: Huh?

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            } catch (IOException e) {
+                logger.error(UNIVERSAL_MESSAGE, e);
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private void messageOfInternalError(@NotNull final HttpSession session,
+                                        @NotNull final String message) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:247]: Какое-то необычное использование параметра -- строка, которая будет выведена в лог, если произойдет ошибка.

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+     */
+    @Path("/v0/entity")
+    @RequestMethod(Request.METHOD_DELETE)
+    public void delete(@Param(required = true, value = "id") @NotNull final String id,
+                       @NotNull final HttpSession session,
+                       @NotNull final Request request) {
+        if (checkId(id, session)) {
+            return;
+        }
+        executor.execute(() -> {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:205]: Здесь может прилететь `RejectedExecutionException` из-за ограниченной очереди, и клиент не получит никакого ответа.

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    public void get(@Param(required = true, value = "id") @NotNull final String id,
+                    @NotNull final HttpSession session,
+                    @NotNull final Request request) {
+        if (checkId(id, session)) {
+            return;
+        }
+        executor.execute(() -> {
+            final ByteBuffer key = ByteBuffer.wrap(id.getBytes(Charsets.UTF_8));
+            final String node = topology.primaryFor(key);
+            if (topology.isMe(node)) try {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:140]: Ой, какое-то не джавовое форматирование. Запустить реформат кода, пожалуйста.

File: src/main/java/ru/mail/polis/service/valaubr/HttpService.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    @Path("/v0/entity")
+    @RequestMethod(Request.METHOD_PUT)
+    public void put(@Param(required = true, value = "id") @NotNull final String id,
+                    @NotNull final Request request, @NotNull final HttpSession session) {
+        if (checkId(id, session)) {
+            return;
+        }
+        executor.execute(() -> {
+            final String node = topology.primaryFor(ByteBuffer.wrap(id.getBytes(Charsets.UTF_8)));
+            if (topology.isMe(node)) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/valaubr/HttpService.java:175]: Не кажется, что логика с `checkId`/`execute`/`isMe`/`proxy` дублируется три раза? Может можно это отрефакторить как-нибудь?

**Review by vaddya — CHANGES_REQUESTED** [review]
К сожалению, много замечаний по реализации. Поздно увидел, что ревью прошлых этапов еще не закончено, так бы отложил свое ревью, конечно.

**Comment by vaddya** [comment]
Закрываю, частично принято с серьезными замечаниями.

**Pull request closed without merging**